### PR TITLE
Porting UFS-WM Noah-MP GSWP3 baseline to land DA workflow

### PR DESCRIPTION
**Pull request by rhaesung**
## Description
This adds an option to use the current UFS-WM land gswp3 baseline in the land DA system. Currently, a single-day DA cycle experiment is available for this setting. The previous offline Noah-MP setting for era5 is still existing.  

## To reproduce
1. git clone -b feature/porting --recursive https://github.com/ufs-community/land-DA_workflow.git
2. ln -fs /scratch2/NAGAPE/epic/UFS_Land-DA/inputs
3. cd land-DA_workflow/
4. module use modulefiles/ && module load landda_hera.intel
5. mkdir build && cd build/
6. ecbuild .. && make -j8
7. cd ..
8. sed -i 's/--cpus-per-task=1/--cpus-per-task=4/g' submit_cycle.sh
9. ./do_submit_cycle.sh settings_DA_cycle_gswp3

### Anticipated changes to regression tests:
- [X] Is baseline change expected ? 

## Subcomponents involved:
- [X] DA_update
- [X] ufs-land-driver
- [X] vector2tile
- [ ] none

## Linked PR's and Issues:
https://github.com/ufs-community/land-DA_workflow/issues/31
https://github.com/ufs-community/land-DA_workflow/issues/8

### Testing (for CM's):
- RDHPCS
    - [X] Hera
    - [X] Orion
    - [ ] Jet
    - [ ] Gaea
    - [ ] Cheyenne
- CI
  - [ ] Completed
- PW-Clouds
  - [ ] AWS
  - [ ] AZURE
  - [ ] GCP

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -44,17 +44,17 @@ include_directories( ${NETCDF_INCLUDE_DIRS} )
...
-  GIT_REPOSITORY https://github.com/ufs-community/ufs-weather-model
-  GIT_SUBMODULES_RECURSE TRUE
-  GIT_TAG develop
-  SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/ufs-weather-model
-  CMAKE_ARGS ${SOURCE_DIR} -DAPP=LND -DMPI=ON -DCMAKE_BUILD_TYPE=Release
-  INSTALL_COMMAND cmake
-  )
-
-add_subdirectory( vector2tile )
+#ExternalProject_Add(ufs-weather-model
```
Comment (jkbk2004) [inline CMakeLists.txt]: Is there a reason to turn off weather model build?

File: tile2tile/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+Code to map between vector format used by the Noah-MP offline driver, and the tile format used by the UFS atmospheric model. Currently used to prepare input tile files for JEDI. These files include only those fields required by JEDI, rather than the full restart. Can also be used to map stochastic_physics output to the tile or vector. Stochastic physics output files depend on the layout, with 1x1 layout giving one file per tile.
```
Comment (jkbk2004) [inline tile2tile/README.md]: Is this description valid for tile2tile?

File: vector2tile/namelist.vector2tile [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+!------------------- common -------------------
+! Direction of conversion: either "vector2tile" or "tile2vector" for restart file
+!                                 "lndp2tile" or "lndp2vector" for perturbation 
+
+  direction = "vector2tile"
+
+! FV3 resolution and path to oro files for restart/perturbation conversion
+
+  tile_size = 96
+  tile_path = "/scratch1/NCEPDEV/stmp2/Michael.Barlage/models/vector/v2t_data/tile_files/C96.mx100_frac/"
```
Comment (jkbk2004) [inline vector2tile/namelist.vector2tile:12]: How is the machine specific path going to be resolved across other machines\: orion, clouds, etc?

**Comment by jkbk2004** [comment]
@rhaesung How do you coordinate to reflect this code change to documentation side?

File: submit_cycle.sh [hunk 4]
```diff
@@ -1,10 +1,10 @@
 #!/bin/bash  
-#SBATCH --job-name=offline_noahmp
-#SBATCH --account=da-cpu
+#SBATCH --job-name=ufs_land_da
+#SBATCH --account=epic
 #SBATCH --qos=debug
 #SBATCH --nodes=1
 #SBATCH --tasks-per-node=6
-#SBATCH --cpus-per-task=1
+#SBATCH --cpus-per-task=5
```
Comment (jkbk2004) [inline submit_cycle.sh]: Is there any impact if era5 option is still available to run? Do users need to change any portion of job card?

**Comment by rhaesung** [comment]
> @rhaesung How do you coordinate to reflect this code change to documentation side?

@jkbk2004 We have a meeting with stakeholders and the US team tomorrow. We will discuss updating the documentation as well. FYI @gspetro-NOAA

**Comment by rhaesung** [comment]
@jkbk2004 I just tested and found that at least '--cpus-per-tasks'=4 is needed to run this experiment without crashing.

**Comment by jkbk2004** [comment]
> @jkbk2004 I just tested and found that at least '--cpus-per-tasks'=4 is needed to run this experiment without crashing.

what about [land-DA_workflow](https://github.com/ufs-community/land-DA_workflow/tree/develop)/[test](https://github.com/ufs-community/land-DA_workflow/tree/develop/test)
/run_ufs_datm_lnd.sh ? ctest bascially 6 cpus, right? am I right? If different resources are needed for era5/gswp, I am not sure it's good idea to keep same job card.

**Comment by rhaesung** [comment]
> > @jkbk2004 I just tested and found that at least '--cpus-per-tasks'=4 is needed to run this experiment without crashing.
> 
> what about [land-DA_workflow](https://github.com/ufs-community/land-DA_workflow/tree/develop)/[test](https://github.com/ufs-community/land-DA_workflow/tree/develop/test) /run_ufs_datm_lnd.sh ? ctest bascially 6 cpus, right? am I right? If different resources are needed for era5/gswp, I am not sure it's good idea to keep same job card.

You are right. ctest is using the datm_cdeps_lnd_gswp3 not datm_cdeps_lnd_gswp3_rst baseline for test purposes.

**Comment by jkbk2004** [comment]
> > > @jkbk2004 I just tested and found that at least '--cpus-per-tasks'=4 is needed to run this experiment without crashing.
> > 
> > 
> > what about [land-DA_workflow](https://github.com/ufs-community/land-DA_workflow/tree/develop)/[test](https://github.com/ufs-community/land-DA_workflow/tree/develop/test) /run_ufs_datm_lnd.sh ? ctest bascially 6 cpus, right? am I right? If different resources are needed for era5/gswp, I am not sure it's good idea to keep same job card.
> 
> You are right. ctest is using the datm_cdeps_lnd_gswp3 not datm_cdeps_lnd_gswp3_rst baseline for test purposes.

reduced resources, right?

**Comment by rhaesung** [comment]
> > > > @jkbk2004 I just tested and found that at least '--cpus-per-tasks'=4 is needed to run this experiment without crashing.
> > > 
> > > 
> > > what about [land-DA_workflow](https://github.com/ufs-community/land-DA_workflow/tree/develop)/[test](https://github.com/ufs-community/land-DA_workflow/tree/develop/test) /run_ufs_datm_lnd.sh ? ctest bascially 6 cpus, right? am I right? If different resources are needed for era5/gswp, I am not sure it's good idea to keep same job card.
> > 
> > 
> > You are right. ctest is using the datm_cdeps_lnd_gswp3 not datm_cdeps_lnd_gswp3_rst baseline for test purposes.
> 
> reduced resources, right?

era5/ctest need 6 cpus and gswp3 needs 24 cpus. do we want a separate job card for two different settings?

**Comment by jkbk2004** [comment]
I still believe something more can be done when you set up for gswp. But since we have no time, separate job card for gswp might be a safer way to keep era5 feature intact. Clearly, we need to clean up the issue later.

**Comment by jkbk2004** [comment]
We call sbatch slurm script as job card (in this case, submit_cycle.sh). If you can specifically separate out different script for gswp, that's one way to think.

**Comment by rhaesung** [comment]
> We call sbatch slurm script as job card (in this case, submit_cycle.sh). If you can specifically separate out different script for gswp, that's one way to think.

Thanks, @jkbk2004 !

**Comment by jkbk2004** [comment]
> We call sbatch slurm script as job card (in this case, submit_cycle.sh). If you can specifically separate out different script for gswp, that's one way to think.

@rhaesung if you feel, it's time consuming to separate. Then, we can think to keep develop feature of era5 by default and specify gswp job card configuration in the document.

**Comment by rhaesung** [comment]
> > We call sbatch slurm script as job card (in this case, submit_cycle.sh). If you can specifically separate out different script for gswp, that's one way to think.
> 
> @rhaesung if you feel, it's time consuming to separate. Then, we can think to keep develop feature of era5 by default and specify gswp job card configuration in the document.

Great, I like this idea!

**Comment by jkbk2004** [comment]
@rhaesung I didn't mean to totally revert the change in the script. You need to keep gswp flags in the script if people try to test with gswp. I mean only to set sbatch resource configuration for era5 by default

**Comment by rhaesung** [comment]
> @rhaesung I didn't mean to totally revert the change in the script. You need to keep gswp flags in the script if people try to test with gswp. I mean only to set sbatch resource configuration for era5 by default

yes, I understood.

**Comment by jkbk2004** [comment]
@ulmononian note we are about to merge this pr.

**Comment by ulmononian** [comment]
@jkbk2004 ok so new container build needed for develop once this is merged? with minor instructions to run workflow?

**Comment by jkbk2004** [comment]
@ulmononian we are adding gswp case thru [settings_DA_cycle_gswp3](https://github.com/ufs-community/land-DA_workflow/blob/feature/porting/settings_DA_cycle_gswp3), which is based on cubed sphere native grid (no conversion btw vetor2tile is needed). But still quite some code cleanup process is needed and we don't have much time. So we will go ahead to keep current develop branch feature of existing era5 test. One more round of just container update should be ok after merge.  It sounds like we will create tag for v1.2 after thanksgiving.

**Comment by jkbk2004** [comment]
@gspetro-NOAA FYI

**Comment by jkbk2004** [comment]
@ulmononian BTW, orion module update might be needed. If SI can combine orion update to this pr, we can hold the PR 
a bit more. Or separate PR

**Comment by rhaesung** [comment]
> @ulmononian BTW, orion module update might be needed. If SI can combine orion update to this pr, we can hold the PR a bit more. Or separate PR

@natalie-perlin is also aware of this issue: https://github.com/ufs-community/land-DA_workflow/issues/35.

**Comment by zach1221** [comment]
@rhaesung pipelines ctests pass on hera.

Edit: log file too large to post. Attaching link to the pipeline console output instead. [Here](https://jenkins.epic.oarcloud.noaa.gov/job/land-da/job/Land%20DA%20Jenkinsfile/view/change-requests/job/PR-32/17/console)

**Comment by ulmononian** [comment]
@rhaesung @jkbk2004 what is the version of spack-stack you tested and expected to use for this PR? spack-stack/1.3.0?

**Comment by rhaesung** [comment]
> @rhaesung pipelines ctests pass on hera. [Land-DA PR 32 Test log.pdf](https://github.com/ufs-community/land-DA_workflow/files/13368739/Land-DA.PR.32.Test.log.pdf)

@zach1221 Great, thanks for confirming.

**Comment by rhaesung** [comment]
> @rhaesung @jkbk2004 what is the version of spack-stack you tested and expected to use for this PR? spack-stack/1.3.0?

@ulmononian spack-stack/1.3.0 is working fine on hera.

**Comment by ulmononian** [comment]
@rhaesung two options: i can try to install spack-stack/1.3.0 in the new space on orion, or you can test on orion with spack-stack/1.5.0 or spack-stack/1.5.1 (but this will include some library version changes). do you have a preference?

**Comment by jkbk2004** [comment]
@ulmononian note that weather model develop advanced to spack stack 1.5. Somehow this pr (and even develop branch) build ok with weather model and spack stack 1.3: https://github.com/ufs-community/land-DA_workflow/blob/feature/porting/modulefiles/landda_hera.intel.lua. Certainly
, we should sync up the version. But I don't want to add more urgency.

**Comment by rhaesung** [comment]
> @ulmononian note that weather model develop advanced to spack stack 1.5. Somehow this pr (and even develop branch) build ok with weather model and spack stack 1.3: https://github.com/ufs-community/land-DA_workflow/blob/feature/porting/modulefiles/landda_hera.intel.lua. Certainly , we should sync up the version. But I don't want to add more urgency.

@ulmononian I agree.

**Comment by ulmononian** [comment]
10-4. installing 1.3.0 on orion now. will let you know when done.

**Comment by ulmononian** [comment]
spack-stack/1.3.0 is back on orion in ` /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.3.0/envs/unified-env`

**Comment by rhaesung** [comment]
@ulmononian thanks! should I just change the path of spack-stack in https://github.com/ufs-community/land-DA_workflow/blob/feature/porting/modulefiles/landda_orion.intel.lua?

**Comment by ulmononian** [comment]
> @ulmononian thanks! should I just change the path of spack-stack in https://github.com/ufs-community/land-DA_workflow/blob/feature/porting/modulefiles/landda_orion.intel.lua?

@rhaesung yes that should do it

**Comment by rhaesung** [comment]
> > @ulmononian thanks! should I just change the path of spack-stack in https://github.com/ufs-community/land-DA_workflow/blob/feature/porting/modulefiles/landda_orion.intel.lua?
> 
> @rhaesung yes that should do it

@ulmononian can you please try? looks like it's keep loading

**Comment by ulmononian** [comment]
please change netcdf-c to `netcdf-c/4.9.2`. all else is identical.

**Comment by rhaesung** [comment]
> please change netcdf-c to `netcdf-c/4.9.2`. all else is identical.

@ulmononian I still see the below error:

land-DA_workflow$module load landda_orion.intel 
^CLmod has detected the following error:  Unable to load module because of error when
evaluating modulefile:
     modulefiles/landda_orion.intel.lua: /apps/lmod/lmod/libexec/DirTree.lua:100: interrupted!
     Please check the modulefile and especially if there is a the line number specified in the
above message 
While processing the following module(s):
    Module fullname     Module Filename
    ---------------     ---------------
    landda_orion.intel  modulefiles/landda_orion.intel.lua

**Comment by rhaesung** [comment]
Hi @ulmononian, any update in  [https://github.com/ufs-community/land-DA_workflow/blob/feature/porting/modulefiles/landda_orion.intel.lua](https://github.com/ufs-community/land-DA_workflow/blob/feature/porting/modulefiles/landda_orion.intel.lua)?

**Comment by jkbk2004** [comment]
@rhaesung I don't understand. I am able to build ok with /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.3.0/envs/unified-env/install/modulefiles/Core. I will push new orion modulepath @ulmononian @zach1221 FYI

**Comment by jkbk2004** [comment]
> @rhaesung I don't understand. I am able to build ok with /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.3.0/envs/unified-env/install/modulefiles/Core. I will push new orion modulepath @ulmononian @zach1221 FYI

BTW, my setup is at /work/noaa/epic/jongkim/landv1.2/land-DA_workflow

**Comment by jkbk2004** [comment]
@rhaesung why are you creating all kinds of feature branches in this authoritative repo?

**Comment by rhaesung** [comment]
> @rhaesung I don't understand. I am able to build ok with /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.3.0/envs/unified-env/install/modulefiles/Core. I will push new orion modulepath @ulmononian @zach1221 FYI

@jkbk2004 It's interesting, this didn't work last week on my side. Thanks for checking and pushing the new modulepath.

**Comment by rhaesung** [comment]
> @rhaesung why are you creating all kinds of feature branches in this authoritative repo?

I will clean them up.

**Comment by rhaesung** [comment]
@jkbk2004 @ulmononian Testing this on Orion and got this error when running the JEDI:

fv3jedi_letkf.x: error while loading shared libraries: libatlas_f.so.0.32: cannot open shared object file: No such file or directory

Looking into this and I suspect it might be needed to reinstall skylab v4.0 with spack-stack 1.3.0 if possible. FYI @zach1221

**Comment by ulmononian** [comment]
@jkbk2004 so new 1.3.0 orion stack is working ok?

**Comment by jkbk2004** [comment]
> @jkbk2004 so new 1.3.0 orion stack is working ok?

@ulmononian I think we need to do jedi ctests with skylab build to address the atlas issue reported, Can we re-install skylab-v4.0? Or let me know if you want me to check.

**Comment by rhaesung** [comment]
> > @jkbk2004 so new 1.3.0 orion stack is working ok?
> 
> @ulmononian I think we need to do jedi ctests with skylab build to address the atlas issue reported, Can we re-install skylab-v4.0? Or let me know if you want me to check.

@jkbk2004 @ulmononian I've tried to re-install the skylabv4.0 w spack-stack 1.3.0 and only the 'ewok-env' module didn't work.

**Comment by jkbk2004** [comment]
jedi-bundle ctest looks good: 99% tests passed, 2 tests failed out of 1229. I am not concern with two failures. @rhaesung installation is available at /work/noaa/epic/UFS_Land-DA-v1.2/

**Comment by rhaesung** [comment]
> jedi-bundle ctest looks good: 99% tests passed, 2 tests failed out of 1229. I am not concern with two failures. @rhaesung installation is available at /work/noaa/epic/UFS_Land-DA-v1.2/

Thanks @jkbk2004 for the installation. Testing again now!

**Comment by rhaesung** [comment]
@jkbk2004 @zach1221 All tests are passed. Please let me know if you have any issues.

**Comment by zach1221** [comment]
> @jkbk2004 @zach1221 All tests are passed. Please let me know if you have any issues.

Hera is still good running via jenkins. Just checking orion once more.

**Comment by zach1221** [comment]
Ok apologies for the delay. The jenkins pipeline now runs on both hera and orion, with all ctests passing. 
Console output can be viewed [here](https://jenkins.epic.oarcloud.noaa.gov/job/land-da/job/Land%20DA%20Jenkinsfile/view/change-requests/job/PR-32/40/).

**Review by jkbk2004 — DISMISSED** [review]

File: tile2tile/README.md (hunk 2 above)
Comment (rhaesung) [reply tile2tile/README.md]: README is removed.

File: vector2tile/namelist.vector2tile (hunk 3 above)
Comment (rhaesung) [reply vector2tile/namelist.vector2tile:12]: This is just a template. We are actually using the land-DA_workflow/template.vector2tile file.

**Review by jkbk2004 — APPROVED** [review]

**Pull request merged**